OSDOCS#6518: 64k pages in the RHCOS kernal support on multi-arch compute (#69212)
openshift/openshift-docs: modules/multi-architecture-enabling-64k-pages.adoc

//Module included in the following assemblies
//
//post_installation_configuration/multi-architecture-configuration.adoc

:_mod-docs-content-type: PROCEDURE
[id="multi-architecture-enabling-64k-pages_{context}"]

= Enabling 64k pages on the {op-system-first} kernel 

> You can enable 64k memory pages in the {op-system-first} kernel on the 64-bit ARM compute machines in your cluster. The 64k page kernel specification can be used for large GPU or high memory workloads. This is done using the Machine Config Operator (MCO) which updates the {op-system} kernel. You can create a custom machine config pool to add 64k pages to your nodes. 

Reviewer: "This is done by the Machine Config Operator (MCO) using machine config pools which updates the {op-system} kernel" ?

Author: Maybe I can word that better. I thought the MCO is the operator that updates the kernel. Like how we would make a machinepool and the MCO will use the machinepool to update the kernel. Or is that not accurate?

Reviewer: correct. so the MCO uses the machineconfigpool to update  the kernel.

> You can enable 64k memory pages in the {op-system-first} kernel on the 64-bit ARM compute machines in your cluster. The 64k page kernel specification can be used for large GPU or high memory workloads. This is done using the Machine Config Operator (MCO) which updates the {op-system} kernel. You can create a custom machine config pool to add 64k pages to your nodes. 

Reviewer 2: > 64k memory pages

I wonder if we should call out the feature as `64k pagesize kernel` throughout the doc. Although it's shorter and clear enough to call it out as "64k pages", I'd keep some more formality/precision in the documentation.

Author: I've been using `64k pages` as the main feature name, but in introductory sentences or in the beginning of some procedures I've been using things like `64k page kernel` or `64k memory pages` just to add more details when introducing the topic.

Reviewer 2: Should we use `64k page size kernel` as main feature name based on the discussion in the [other PR](https://github.com/openshift/openshift-docs/pull/69875#discussion_r1447719960) and for consistency with the RHEL doc?

[IMPORTANT]
====

> Using 64k pages is exclusive to 64-bit ARM architecture compute nodes or clusters installed on 64-bit ARM machines. If you create a machine pool on a 64-bit x86 built cluster, the machine pool and operator will degrade.

Author: Great suggestion, thank you!

====

.Prerequisites: 
* You installed the OpenShift CLI `oc`
* You created a cluster with compute nodes of different architecture on one of the supported platforms

.Procedure: 

. Create the machine config pool on your compute node by running the following 
command:
+
[source,terminal]
----
$ oc create -f <filename>.yaml
----

> . Edit the machine config pool and add the `64k-pages` parameter to the configuration:

Reviewer: we need to recommend that the user creates a separate machine config pool for arm64 nodes and apply the config to the pool. [This ](https://access.redhat.com/solutions/5688941)article details how to create a machineconfigpool

Reviewer 2: Also, we can use the procedure at https://docs.openshift.com/container-platform/4.13/machine_management/creating-infrastructure-machinesets.html#creating-infra-machines_creating-infrastructure-machinesets, except the optional point (5) that does not apply for us.

Example MachineConfigPool that fits our case:

```yaml
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfigPool
metadata:
  name: worker-64k-pages
spec:
  machineConfigSelector:
    matchExpressions:
      - key: machineconfiguration.openshift.io/role
        operator: In
        values:
        - worker
        - worker-64k-pages
  nodeSelector:
    matchLabels:
      node-role.kubernetes.io/worker-64k-pages: ""
      kubernetes.io/arch: arm64
```

//Module included in the following assemblies
//
//post_installation_configuration/multi-architecture-configuration.adoc

:_mod-docs-content-type: PROCEDURE
[id="multi-architecture-enabling-64k-pages_{context}"]

= Enabling 64k pages on the {op-system-first} kernel 

> You can enable 64k memory page in the {op-system-first} kernel on the 64-bit ARM compute machines in your cluster. The 64k page size kernel specification can be used for large GPU or high memory workloads. This is done using the Machine Config Operator (MCO) which uses a machine config pool to update the kernel To enable 64k page sizes, you must create a machine config pool for AMD64 and a machine config pool for ARM64.

Reviewer 2: ```suggestion
You can enable 64k memory page in the {op-system-first} kernel on the 64-bit ARM compute machines in your cluster. The 64k page size kernel specification can be used for large GPU or high memory workloads. This is done using the Machine Config Operator (MCO) which uses a machine config pool to update the kernel To enable 64k page sizes, you must dedicate a machine config pool for the ARM64 machines to enable this kernel on.
```

We don't mandatory need another machine config pool for the amd64 machines. We must guarantee a dedicated machine config pool for those arm64 machine that should run the 64k pagesize kernel.

Author: Okay gotcha, that makes sense to me!

[IMPORTANT]
====
Using 64k pages is exclusive to 64-bit ARM architecture compute nodes or clusters installed on 64-bit ARM machines. If you configure the 64k pages kernel on a machine config pool using 64-bit x86 machines, the machine config pool and operator will degrade.
====

.Prerequisites: 
* You installed the OpenShift CLI `oc`
* You created a cluster with compute nodes of different architecture on one of the supported platforms

.Procedure: 

. Add a label to the node you want to assign as the infra node with a specific label: 
//not sure if this part was needed
[source,terminal]
+
----

> $ oc label node <node_name> <label>

Reviewer 2: should we specify the example label here?

```
node-role.kubernetes.io/worker-64k-pages: ""
```

Yes, this part is needed too.

I'm not sure whether to specify that in some cases (IPI clusters using the MachineApi), users can also define this label at MachineSet level.

Author: I think i understand your suggestion, lemme add that!

----

> . Create a machine config pool that contains the worker role that uses the ARM64 architecture and the `64k-pages` role:

Reviewer 2: ```suggestion
. Create a machine config pool that contains the worker role that uses the ARM64 architecture and the `worker-64k-pages` role:
```

[source,yaml]
+
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfigPool
metadata:
  name: worker-64k-pages
spec:
  machineConfigSelector:
    matchExpressions:
      - key: machineconfiguration.openshift.io/role
        operator: In
        values:

>         - worker

Reviewer: should we advise to patch this config to remove the  `worker` role? i assume it wouldn't include the amd64 nodes as well because of the arch label selector?

Author: Hmm, I'm not too sure, @aleskandro Do you know this one?

Reviewer 2: I'm not sure I got you 100% @Prashanth684 .

Users should not remove the `worker` element in the MachineConfig's selector's value: the child MCP `worker-64k-pages` should still make use of the MachineConfigurations bound to the built-in `worker` role and add the ones bound to the custom role we are documenting.

Instead, yes, the nodeSelector will take care of binding only arm64 nodes to the custom MCP

Reviewer: ok..my concern was that it would not filter correctly if all nodes in the worker pool was also included, but looks like that's not the case so all good.

//Module included in the following assemblies
//
//post_installation_configuration/multi-architecture-configuration.adoc

:_mod-docs-content-type: PROCEDURE
[id="multi-architecture-enabling-64k-pages_{context}"]

= Enabling 64k pages on the {op-system-first} kernel 

> You can enable 64k memory page in the {op-system-first} kernel on the 64-bit ARM compute machines in your cluster. The 64k page size kernel specification can be used for large GPU or high memory workloads. This is done using the Machine Config Operator (MCO) which uses a machine config pool to update the kernel. To enable 64k page sizes, you must dedicate a machine config pool for ARM64 to enable on the kernel.

Reviewer 3: ```suggestion
You can enable the 64k memory page in the {op-system-first} kernel on the 64-bit ARM compute machines in your cluster. The 64k page size kernel specification can be used for large GPU or high memory workloads. This is done using the Machine Config Operator (MCO) which uses a machine config pool to update the kernel. To enable 64k page sizes, you must dedicate a machine config pool for ARM64 to enable on the kernel.
```

[IMPORTANT]
====

> Using 64k pages is exclusive to 64-bit ARM architecture compute nodes or clusters installed on 64-bit ARM machines. If you configure the 64k pages kernel on a machine config pool using 64-bit x86 machines, the machine config pool and operator will degrade.

Reviewer 3: ```suggestion
Using 64k pages is exclusive to 64-bit ARM architecture compute nodes or clusters installed on 64-bit ARM machines. If you configure the 64k pages kernel on a machine config pool using 64-bit x86 machines, the machine config pool and Machine Config Operator will degrade.
```

====

.Prerequisites: 

> * You installed the OpenShift CLI `oc`

Reviewer 3: ```suggestion
* You installed the OpenShift CLI (`oc`).
```

> * You created a cluster with compute nodes of different architecture on one of the supported platforms

Reviewer 3: ```suggestion
* You created a cluster with compute nodes of different architecture on one of the supported platforms.
```

.Procedure: 

. Label the nodes where you want to run the 64k page size kernel:
[source,terminal]
+
----
$ oc label node <node_name> <label>
----
+

> [source,terminal]

Reviewer 3: ```suggestion
.Example command
[source,terminal]
```

----
$ oc label node worker-arm64-01 node-role.kubernetes.io/worker-64k-pages=
----

> 

Reviewer 3: Actually, I don't think we need both a generic command and a command with example values. I might go with the latter, so that the user can see the label being used in the MCP.

Author: I believe QE wanted to show the generic command and we thought having an example would just be helpful for users.

. Create a machine config pool that contains the worker role that uses the ARM64 architecture and the `worker-64k-pages` role:
[source,yaml]
+
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfigPool
metadata:
  name: worker-64k-pages
spec:
  machineConfigSelector:
    matchExpressions:
      - key: machineconfiguration.openshift.io/role
        operator: In
        values:
        - worker
        - worker-64k-pages
  nodeSelector:
    matchLabels:
      node-role.kubernetes.io/worker-64k-pages: ""
      kubernetes.io/arch: arm64
----

> . You need to also create a machine config on your compute node to enable `64k-pages`

Reviewer 3: Or, should this be:
```suggestion
. Create the object.
```
I'm not sure what we are creating here, as the next step is to edit the machine config (which doesn't exist yet, if I follow this right).

> . You need to also create a machine config on your compute node to enable `64k-pages`

Author: Its mostly just separating the command, so you make the file then edit the file and add the example MachineConfig shown. I rephrased the edit command cuz I see that is unclear.  
I know there is a way to edit the machine config by just running something like
oc create -f <file> <the parameters for the MachineConfig> 
but we thought this way was easier to follow.

> . You need to also create a machine config on your compute node to enable `64k-pages`

Reviewer 2: Yes, we'd create the object from a yaml, where the yaml is specified later (no need for the edit part).  Thanks @mburke5678

> . You need to also create a machine config on your compute node to enable `64k-pages`

Author: Gotcha! I misunderstood the comment at first, thank you @aleskandro & @mburke5678

+
[source,terminal]
----
$ oc create -f <filename>.yaml
----

> . Edit the machine config and add the `64k-pages` parameter to the configuration:

Reviewer 3: ```suggestion
. Create a machine config for the new machine config pool and add the `64k-pages` parameter to the configuration:
```

+

> [source,terminal]

Reviewer 3: I don't think we need this command here.

----
$ oc edit <filename>.yaml
----
+
.Example MachineConfig
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1

> kind: MachineConfigPool

Reviewer 4: The kind is "MachineConfigPool", but it should be "MachineConfig", shoudn't it?

Author: Hmm, I thought it was intentionally MachineConfig, @aleskandro or @Prashanth684 do you know?

metadata:
  labels:

>     machineconfiguration.openshift.io/role: "worker" <1>

Reviewer 4: To make this MachineConfig consistent with the other resources in this doc (the pool and the node label) we should use `machineconfiguration.openshift.io/role:  "worker-64k-pages"` instead.

The current example is a machineconfig that will enable the 64k-pages kernel in the "worker" pool and not in the "worker-64-pages" pool.

The comment <1> tells the customer to edit it if necessary, but wouldn't it be better to use an example value that is consistent with the rest of the resources? I'm afraid of customers copy pasting the resources and finding out that they are applying the kernel to the wrong pool.

>     machineconfiguration.openshift.io/role: "worker" <1>

Reviewer 4: To make this MachineConfig consistent with the other resources in this doc (the pool and the node label) we should use machineconfiguration.openshift.io/role:  "worker-64k-pages" instead.

The current example is a machineconfig that will enable the 64k-pages kernel in the "worker" pool and not in the "worker-64-pages" pool.

The comment <1> tells the customer to edit it if necessary, but wouldn't it be better to use an example value that is consistent with the rest of the resources? I'm afraid of customers copy-pasting the resources and finding out that they are applying the kernel to the wrong pool.

> spec:

Reviewer 4: It looks like it is not properly indented. It should be like this:
```
spec:
  kernelType: 64k-pages

```

kernelType: 

>     64k-pages <2>

Reviewer 4: It looks like it is not properly indented. It should be like this:
```
spec:
  kernelType: 64k-pages

```

----

> <1> Specify the value of the `machineconfiguration.openshift.io/role` label in the custom machine config pool. Only update if you are not using the built-in worker pool exhibited in the example MachineConfig

Reviewer 2: We can probably get rid of this by defaulting to the custom MCP:
```suggestion
```

Author: I just updated the text a bit, WDYT @aleskandro

Reviewer 2: sgtm

<2> Specify your desired kernel type. Valid values are `64k-pages` and `default`
+
[NOTE]
====

> The `realtime` type is not supported on 64-bit ARM architecture based compute nodes. The `64k-pages` type is not supported on 64-bit x86 architecture based compute nodes. 

Reviewer 2: ```suggestion
The `realtime` type is supported only on 64-bit x86 architecture based compute nodes. The `64k-pages` type is supported only on 64-bit ARM architecture based compute nodes. 
```

@kelbrown20 should we write it this way to consider P/Z?

Author: Ooo yes, I like this change!

Reviewer: Does P support switching to 4k pages? i think not?

Author: So its okay to leave it as is or should I adjust the wording here?

> ====

Reviewer 3: I think we need this:
```
. Create the object
+
[source,terminal]
----
$ oc create -f <filename>.yaml
----
```

//Module included in the following assemblies
//
//post_installation_configuration/multi-architecture-configuration.adoc

:_mod-docs-content-type: PROCEDURE
[id="multi-architecture-enabling-64k-pages_{context}"]

= Enabling 64k pages on the {op-system-first} kernel 

> You can enable 64k memory page in the {op-system-first} kernel on the 64-bit ARM compute machines in your cluster. The 64k page size kernel specification can be used for large GPU or high memory workloads. This is done using the Machine Config Operator (MCO) which uses a machine config pool to update the kernel To enable 64k page sizes, you must dedicate a machine config pool for ARM64 to enable on the kernel.

Reviewer 2: ```suggestion
You can enable 64k memory page in the {op-system-first} kernel on the 64-bit ARM compute machines in your cluster. The 64k page size kernel specification can be used for large GPU or high memory workloads. This is done using the Machine Config Operator (MCO) which uses a machine config pool to update the kernel. To enable 64k page sizes, you must dedicate a machine config pool for ARM64 to enable on the kernel.
```

Author: Great catch!

[IMPORTANT]
====
Using 64k pages is exclusive to 64-bit ARM architecture compute nodes or clusters installed on 64-bit ARM machines. If you configure the 64k pages kernel on a machine config pool using 64-bit x86 machines, the machine config pool and operator will degrade.
====

.Prerequisites: 
* You installed the OpenShift CLI `oc`
* You created a cluster with compute nodes of different architecture on one of the supported platforms

.Procedure: 

> . Add a label to the node you want to assign as the infra node with a specific label: 

Reviewer 2: ```suggestion
. Label the node(s) where you want to run the 64k pagesize kernel:
```

//not sure if this part was needed
[source,terminal]
+
----
$ oc label node <node_name> <label>
----
+
[source,terminal]
----

> $ oc label node node-role.kubernetes.io/worker-64k-pages: ""

Reviewer 2: We'd add a nodeRef between the `node` resource name and the label, like
```suggestion
$ oc label node worker-arm64-01 node-role.kubernetes.io/worker-64k-pages: ""
```

> $ oc label node worker-arm64-01 node-role.kubernetes.io/worker-64k-pages: ""

Reviewer 2: ```suggestion
$ oc label node worker-arm64-01 node-role.kubernetes.io/worker-64k-pages=
```

kind: MachineConfigPool
metadata:
  name: worker-64k-pages
spec:
  machineConfigSelector:
    matchExpressions:
      - key: machineconfiguration.openshift.io/role
        operator: In
        values:
        - worker
        - worker-64k-pages
  nodeSelector:
    matchLabels:
      node-role.kubernetes.io/worker-64k-pages: ""
      kubernetes.io/arch: arm64
----

. You need to also create a machine config on your compute node to enable `64k-pages`
+
[source,terminal]
----
$ oc create -f <filename>.yaml
----

. Edit the machine config and add the `64k-pages` parameter to the configuration:
+
[source,terminal]
----
$ oc edit <filename>.yaml
----
+
.Example MachineConfig
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfigPool
metadata:
  labels:
    machineconfiguration.openshift.io/role: "worker" <1>
spec:
kernelType: 
    64k-pages <2>
----
<1> Specify the value of the `machineconfiguration.openshift.io/role` label in the custom machine config pool. Only update if you are not using the built-in worker pool exhibited in the example MachineConfig
<2> Specify your desired kernel type. Valid values are `64k-pages` and `default`
+
[NOTE]
====

> The `realtime` type is not supported on 64-bit ARM architecture based clusters and compute nodes. 

Reviewer 2: I'd not highlight `clusters and compute nodes`, `compute nodes` should be enough.

Should we also add
```suggestion
The `realtime` type is not supported on 64-bit ARM architecture based compute nodes. 
The `64k-pages` type is not supported on 64-bit X86 architecture based compute nodes.
```
